packages/astro/src/core/render/core.ts

```diff
@@ -84,65 +109,74 @@ export async function getParamsAndProps(
 	return [params, pageProps];
 }
 
-export async function renderPage(mod: ComponentInstance, ctx: RenderContext, env: Environment) {
-	const paramsAndPropsRes = await getParamsAndProps({
-		logging: env.logging,
-		mod,
-		route: ctx.route,
-		routeCache: env.routeCache,
-		pathname: ctx.pathname,
-		ssr: env.ssr,
-	});
+export async function renderPageWithParamsAndProps() {}
 
-	if (paramsAndPropsRes === GetParamsAndPropsError.NoMatchingStaticPath) {
-		throw new AstroError({
-			...AstroErrorData.NoMatchingStaticPathFound,
-			message: AstroErrorData.NoMatchingStaticPathFound.message(ctx.pathname),
-			hint: ctx.route?.component
-				? AstroErrorData.NoMatchingStaticPathFound.hint([ctx.route?.component])
-				: '',
-		});
-	}
-	const [params, pageProps] = paramsAndPropsRes;
+export type RenderPage = {
+	mod: ComponentInstance;
+	renderContext: RenderContext;
+	env: Environment;
+	apiContext?: APIContext;
+	props: Props;
+	params: Params;
+};
 
+export async function renderPage({
+	mod,
+	renderContext,
+	env,
+	apiContext,
+	params,
+	props,
```

Comment:
can you check and see if it would be possible for `params` and `props` to go on the `renderContext` object? That way there are fewer things that need to be passed to ever function. Conceptually RenderContext is for things specific to a single render, which params and props fits.

I think that renderContext is created *before* params and props, but they could be initialized as empty objects and then be replaced when params and props are gotten.

Reply:
It seemed that `props` and `params` **can** be created when the `renderContext` is created. We just needed the `mod` and the `env`, which are information we already have and are needed for the rendering phase.

My last commit shows the refactor. There's still one more opportunity for another refactor, but we can do it later.